File: leet/0784_letter_case_permutation/0784_letter_case_permutation.py
# LeetCode 784. Letter Case Permutation
# https://leetcode.com/problems/letter-case-permutation/
# O(2^n) time, O(n) space
class Solution:
    def letterCasePermutation(self, S: str) -> List[str]:
        self.answers = []
        self.backtrack(list(S), 0)
        return self.answers

    def backtrack(self, string, i):
        if i == len(string):
            self.answers.append("".join(string))
            return
        string[i] = string[i].upper()
        self.backtrack(string, i + 1)
        if string[i].upper() != string[i].lower():
            string[i] = string[i].lower()
            self.backtrack(string, i + 1)

File: leet/0784_letter_case_permutation/0784_letter_case_permutation.ts
// LeetCode 784. Letter Case Permutation
// https://leetcode.com/problems/letter-case-permutation/
// O(2^n) time-and-space
function letterCasePermutation(S: string): string[] {
    function backtrack(str: string[], i: number) {
        if (i == S.length) {
            ret.push(str.join(''));
            return;
        }
        if (str[i] >= '0' && str[i] <= '9')
            backtrack(str, i + 1);
        else {
            str[i] = str[i].toUpperCase();
            backtrack(str, i + 1);
            str[i] = str[i].toLowerCase();
            backtrack(str, i + 1);
        }
    }
    const ret = [];
    backtrack(S.split(''), 0);
    return ret;
};
